### ISSUE TITLE
Do not convert doc comments into `#[doc = "..."]` form unless necessary for macros

### ISSUE THREAD

**Issue by petrochenkov** [rust-lang/rust#60935]
Keep them as tokens in AST, this may partially mitigate the perf impact from turning normal comments into doc comments (https://github.com/rust-lang/rust/pull/60930).

```rust
pub struct Attribute {
    pub id: AttrId,
    pub style: AttrStyle,
    pub path: Path,
    pub tokens: TokenStream,
    pub is_sugared_doc: bool,
    pub span: Span,
}
```
=>
```rust
pub struct Attribute {
    pub kind: AttrKind,
    pub span: Span,
    pub id: AttrId,
}

pub enum AttrKind {
    Normal(Path, TokenStream, AttrStyle),
    Doc(Symbol),
}
```

**Comment by nnethercote** [comment]
Because `TokenStream` contains a `Lrc<Vec<TreeAndJoint>>`, this would avoid two allocations per doc comment line: one for the `Lrc` and one for the `Vec`. That would get a big chunk of the potential saving. (We would still need one symbol interning per line, which involves a TLS access and a hash table.)

**Comment by nnethercote** [comment]
I have a local patch that changes `TokenStream` to a `Lrc<SmallVec<[TreeAndJoint; 2]>>`. It's a clear win, but it only gets rid of one of the allocations, so the approach suggested in this PR has a greater potential.

**Comment by nnethercote** [comment]
I have implemented this idea in #65750. It works well!
